### PR TITLE
collectd: collect some stats from cake & mq qdisc instances

### PR DESCRIPTION
**Pull request by ldir-EDB0**
@hnyman 

Is this getting a bit closer?  I know still have to make fq_codel work ;-)

### CODE REVIEW

**Comment by hnyman** [comment]
Do chown & chmod in init script cause error if the plugin file is not there?  
Should the script first check for the existence of the file?

Or should that action be in the plugin's post-install script or something like that  (or can it actually be, so that it survives flashing?)

File: utils/collectd/files/exec-scripts/sqm_collectd.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	json_load "${jsn}"
+	json_get_var qdisc kind
+
+	case "$qdisc" in
+		cake) handle_cake "$ifc"
+		;;
+		fq_codel) handle_fqcodel "$ifc"
+		;;
+
+		*) echo "Unknown qdisc type"
```
Comment (pprindeville) [inline utils/collectd/files/exec-scripts/sqm_collectd.sh:80]: 1>&2

File: utils/collectd/files/exec-scripts/sqm_collectd.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#Options
+	json_select options
+	json_get_vars bandwidth
+	json_select ".."
+
+	echo "PUTVAL \"$HOSTNAME/sqm-$ifc/qdisc_bytes\" interval=$INTERVAL N:$bytes"
+	echo "PUTVAL \"$HOSTNAME/sqm-$ifc/qdisc_drops\" interval=$INTERVAL N:$drops"
+	echo "PUTVAL \"$HOSTNAME/sqm-$ifc/qdisc_backlog\" interval=$INTERVAL N:$backlog"
+
+#Tins
```
Comment (pprindeville) [inline utils/collectd/files/exec-scripts/sqm_collectd.sh:26]: Not a fan of the comments not matching indent...

Comment (ldir-EDB0) [reply utils/collectd/files/exec-scripts/sqm_collectd.sh]: Done, I think.

**Comment by hnyman** [comment]
Should the cake tin rrd file be named sqmcake (and the tab titled as "SQM cake(tin)") ?
Then the sqm and sqmcake based tabs (and possible sqmfqcodel tab) would be next to each other in LuCI.

(Now cake tin is first and SQM is far away on right side)

**Comment by hnyman** [comment]
@ldir-EDB0 
I compiled the newest version, installed the .ipk with opkg, and it fails to run.

```
...
Sat Apr 25 17:03:13 2020 daemon.err collectd[5536]: Initialization complete, entering read-loop.
Sat Apr 25 17:03:14 2020 daemon.err collectd[5536]: exec plugin: exec_read_one: error = exec plugin: Failed to execute ``/usr/bin/collectd/sqm_collectd.sh'': Permission denied
Sat Apr 25 17:04:14 2020 daemon.err collectd[5536]: exec plugin: exec_read_one: error = exec plugin: Failed to execute ``/usr/bin/collectd/sqm_collectd.sh'': Permission denied
```
I think that reason is that the init script sets too tight permissions for the script.

```
root@router1:~# ls -ld /usr/bin/co*
dr-x------    2 root     root           232 Apr 25 17:00 /usr/bin/collectd
root@router1:~# ls -la /usr/bin/collectd/
dr-x------    2 root     root           232 Apr 25 17:00 .
drwxr-xr-x    1 root     root           232 Apr 25 17:00 ..
-r-x------    1 root     root          2615 Apr 25 16:38 sqm_collectd.sh
```

Changing the init script to chmod 0555 seems to work. (although it seems strange that there is no 7 for root). The ownership does not seem to change.

**Comment by ldir-EDB0** [comment]
> Changing the init script to chmod 0555 seems to work. (although it seems strange that there is no 7 for root). The ownership does not seem to change.

This reads better.... and it actually works!
find /usr/bin/collectd ! \( -user nobody -a -group nogroup \) -exec chown nobody:nogroup '{}' '+'

I plan to spend some time on this today, nudging toward perfection 😂

**Comment by hnyman** [comment]
@jow- 
Any comments on the method of introducing the script to be executed by collectd-mod-exec for getting SQM qdisc data via tc?
(It needs to be run as nobody)

Currently the collectd init script gets changed like this: 
https://github.com/openwrt/packages/pull/11941/files#diff-95f599d121a72be87f474efd6eebb2a4R14

**Comment by hnyman** [comment]
Maybe we should maybe merge the first version soon, I think.  Seems to work well enough.
(this and the rrdtool change)

**Comment by jow-** [comment]
Can we move the `/usr/bin/collectd` directory to `/usr/libexec/collectd` ? The libexec folder is more appropriate for things which are not meant to be executed directly.

**Comment by neheb** [comment]
Requires a PKG_RELEASE bump.

**Comment by ldir-EDB0** [comment]
ping

**Comment by hnyman** [comment]
Your code has worked ok for me, (I have updated it in my builds along your changes), so from my side, feel free to merge if you are ready enough.  
(You have still titled this PR as RFC, so I haven't dared to merge it myself ;-)

And the same for rrdtool thing. The log change seems to work ok.

**Pull request merged**